_posts/2020-02-03-fetch-offset-ties.md

```diff
@@ -0,0 +1,99 @@
+excerpt_separator: <!--more-->
+---
+
+Presto follows the SQL Standard faithfully. We extend it only when it is well justified,
+we strive to never break it and we always prefer the standard way of doing things.
+There was one situation where we stumbled, though. We had a non-standard way of limiting
+query results with `LIMIT n` without implementing the standard way of doing that first.
+We have corrected that, adding ANSI SQL way of limiting query results, discarding initial
+results and -- a hidden gem -- retaining initial results in case of ties.
+
+<!--more-->
+
+# Limiting query results
+
+Probably everyone using relational databases knows the `LIMIT n` syntax for limiting query
+results. It is supported by e.g. MySQL, PostgreSQL and many more SQL engines following
+their example. It is so common that one could think that `LIMIT n` is the standard way
+of limiting the query results.  Let's have a look at how various popular SQL engines
+provide this feature.
+
+* DB2, MySQL, MariaDB, PostgreSQL, Redshift, MemSQL, SQLite and many others provide the `... LIMIT n` syntax.
+* SQL Server provides `SELECT TOP n ...` syntax.
+* Oracle provides `... WHERE ROWNUM <= n` syntax.
+
+And what does the SQL Standard say?
+
+```sql
+SELECT *
+FROM my_table
+FETCH FIRST n ROWS ONLY 
+```
+
+If we look again at the database systems mentioned above, it turns out many of them support the standard
+syntax too: Oracle, DB2, SQL Server and PostgreSQL (although that's not documented currently).
+
+And Presto? Presto has `LIMIT n` support since 2012. In [Presto 310](https://prestosql.io/docs/current/release/release-310.html),
+Presto gained also the `FETCH FIRST n ROWS ONLY` support.
+
+Let's have a look beyond the limits.
+
+# Tie break
+
+Admittedly, `FETCH FIRST n ROWS ONLY` syntax is way more verbose than the short `LIMIT n` syntax Presto
+always supported (and still does). However, it is also more powerful: it allows selecting rows "top n,
+ties included". Consider a case where you want to list top 3 students with highest score on an exam.
+What happens if the 3<sup>rd</sup>, 4<sup>th</sup> and 5<sup>th</sup> persons have equal score? Which
+one should be returned? Instead of getting an arbitrary (and indeterminate) result you can use
+the `FETCH FIRST n ROWS WITH TIES` syntax:
+
+```sql
+SELECT student_name, score
+FROM student s JOIN exam_result e ON s.id = e.student_id
+ORDER BY score
+FETCH FIRST 3 ROWS WITH TIES
+```
+
+The `FETCH FIRST n ROWS WITH TIES` clause retains all rows with equal values of the ordering keys (the `ORDER BY` clause) as
+the last row that would be returned by the `FETCH FIRST n ROWS ONLY` clause (or its equivalent, `LIMIT n`).
+
+# Offset
+
+Per the SQL Standard, the `FETCH FIRST n ROWS ONLY` clause can be prepended with `OFFSET m`, to skip `m` initial rows.
+In such a case, it makes sense to use `FETCH NEXT ...` variant of the clause -- it's allowed with and without `OFFSET`,
+but definitely looks better with that clause.
```

Comment:
skip " -- it's allowed...". It makes the piece harder to read.
There are more variants of the clause not mentioned in this post.
Optionally, you could describe all of them in one place in grammar-like style:
FETCH (FIRST | NEXT) n? (ROW | ROWS) (ONLY | WITH TIES)

Reply:
I will keep this, but will also add ref to docs which describe full grammar